### PR TITLE
task 4

### PR DESCRIPTION
**Pull request by epiharma**
- [x] 1 - Task 4.1 is implemented
- [x] 3 - Task 4.2 is implemented lambda links are provided and returns data
- [x] 4 - Task 4.3 is implemented lambda links are provided and products is stored in DB (call Task 4.2 to see the product)
- [x] 5 - Your own Frontend application is integrated with Product Service (/products API) and products from Product Service are represented on Frontend. Link to a working Frontend application is provided for cross-check reviewer.
- [ ] +1 (All languages) - POST /products lambda functions returns error 400 status code if product data is invalid
- [x] +1 (All languages) - All lambdas return error 500 status code on any error (DB connection, any unhandled error in code)
- [x] +1 (All languages) - All lambdas do console.log for each incoming requests and their arguments
- [ ] +1 (All languages) - Use RDS instance instead fo DynamoDB tables. Do not commit your environment variables in serverless.yml to github!
- [ ] +1 (All languages) - Transaction based creation of product (in case stock creation is failed then related to this stock product is not created and not ready to be used by the end user and vice versa) (https://devcenter.kinvey.com/nodejs/tutorials/bl-transactional-support, https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/transaction-apis.html)

endpoints:
  GET - https://nk7psf2sf0.execute-api.us-east-1.amazonaws.com/dev/products 
  GET - https://nk7psf2sf0.execute-api.us-east-1.amazonaws.com/dev/products/{id}
  POST - https://nk7psf2sf0.execute-api.us-east-1.amazonaws.com/dev/products
fe:
https://d29i895vf10gm2.cloudfront.net/

### CODE REVIEW

**Comment by kkibalnichenko** [comment]
Good job! Mark: 7

File: product-service/db/init-tables.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+};
+
+const fillTables = async () => {
+  products.forEach((product) => {
+    post(generateParamsForProduct(product));
+    post(generateParamsForStocks(product));
+  });
+};
+
+fillTables();
```
Comment (SlavaJSFE) [inline product-service/db/init-tables.ts:43]: Blank line is missing

File: product-service/environments/dynamo-db-table-names.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export const DynamoDbTableNames = {
+  Products: process.env.PRODUCT_TABLE_NAME ?? "products",
+  Stocks: process.env.STOCK_TABLE_NAME ?? "stocks",
+}
```
Comment (SlavaJSFE) [inline product-service/environments/dynamo-db-table-names.ts:4]: Blank line is missing

File: product-service/environments/environment.types.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+  region: string;
+  profile: string;
+  jwtSecret: string;
+  dynamo: DynamoTableInfo[];
+}
+
+interface DynamoTableInfo {
+  endpoint?: string;
+  tableName: string;
+}
```
Comment (SlavaJSFE) [inline product-service/environments/environment.types.ts:12]: Blank line is missing

File: product-service/src/functions/create-product/handler.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import {Product} from "@models/product";
+import {errorMessage, loggers} from "@libs/logger";
+import ProductService from "@services/product.service";
+
+const {LOG, ERROR} = loggers("createProduct");
+
+const createProduct = async (event: APIGatewayEvent) => {
+  try {
+    const payload = event.body as Partial<Product>;
+    LOG(`Creating product ${JSON.stringify(payload)}`);
```
Comment (SlavaJSFE) [inline product-service/src/functions/create-product/handler.ts:13]: Do you really need 12th and 13th lines in try/catch block?

File: product-service/src/functions/get-product-by-id/handler.ts [hunk 5]
```diff
@@ -1,17 +1,27 @@
...
 
-export const getProductById: ValidatedEventAPIGatewayProxyEvent<any> = async (event) => {
-  const id = (event.pathParameters.id).toString();
+const {LOG, ERROR} = loggers("getProductById");
 
-  const product = products.find(product => product.id === id);
-  if (!product) return { statusCode: 404, body: null }
+export const getProductById = async (event: APIGatewayEvent) => {
+  try {
+    const id = String(event.pathParameters.id);
```
Comment (SlavaJSFE) [inline product-service/src/functions/get-product-by-id/handler.ts:11]: Do you really need this in try/catch block?

File: product-service/src/functions/get-product-by-id/index.ts [hunk 6]
```diff
@@ -1,12 +1,12 @@
-import { handlerPath } from '@libs/handler-resolver';
+import {handlerPath} from "@libs/handler-resolver";
```
Comment (SlavaJSFE) [inline product-service/src/functions/get-product-by-id/index.ts:1]: For destructuring spaces are required

File: product-service/src/functions/get-products/index.ts [hunk 7]
```diff
@@ -1,12 +1,12 @@
-import { handlerPath } from '@libs/handler-resolver';
+import {handlerPath} from "@libs/handler-resolver";
```
Comment (SlavaJSFE) [inline product-service/src/functions/get-products/index.ts:1]: For destructuring spaces are required

File: product-service/src/functions/index.ts [hunk 8]
```diff
@@ -1,2 +1,3 @@
-export { default as getProducts } from './get-products';
-export { default as getProductById } from './get-product-by-id';
\ No newline at end of file
+export {default as getProducts} from "./get-products";
+export {default as getProductById} from "./get-product-by-id";
+export {default as createProduct} from "./create-product";
```
Comment (SlavaJSFE) [inline product-service/src/functions/index.ts:3]: Blanc line is missing

File: product-service/src/libs/db.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    return await dynamo
+      .put({
+        TableName,
+        Item,
+      })
+      .promise();
+  }
+}
+
+export default DB;
```
Comment (SlavaJSFE) [inline product-service/src/libs/db.ts:32]: Blanc line is missing

File: product-service/src/libs/logger.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    LOG: logger("log", handler),
+    WARN: logger("warn", handler),
+    ERROR: logger("error", handler),
+  }
+}
+
+export function errorMessage(e: any) {
+  const message = e.message || e.errorMessage || JSON.stringify(e, null, 2)
+  return String(message)
+}
```
Comment (SlavaJSFE) [inline product-service/src/libs/logger.ts:16]: Blanc line is missing

File: product-service/src/models/product.ts [hunk 11]
```diff
@@ -1,7 +1,7 @@
...
-    title: string;
-    description?: string;
-    price: number;
-    count: number;
+  id: string;
+  title: string;
+  description?: string;
+  price: number;
+  count: number;
 }
```
Comment (SlavaJSFE) [inline product-service/src/models/product.ts:7]: Blanc line is missing

File: product-service/src/services/product.service.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    const id = uuid();
+    const product = {id, ...payload} as Product;
+    const productStock = {product_id: id, count: product.count}
+
+    await DB.put(DynamoDbTableNames.Products, product);
+    await DB.put(DynamoDbTableNames.Stocks, productStock);
+
+    return product;
+  }
+}
```
Comment (SlavaJSFE) [inline product-service/src/services/product.service.ts:37]: Blanc line is missing

**Review by SlavaJSFE — APPROVED** [review]

**Comment by SlavaJSFE** [comment]
Create Product lambda creates new product but return 502 error.
All basic requirements are implemented - 5 points
Also, two additional tasks are implemented - 2 points
#### Total task score - 7 points

**Pull request merged**